rasterio/io.py

```diff
@@ -309,6 +319,64 @@ def open(self, path):
         return s
 
 
+class VirtualWarpedFile(object):
+
+    """Creates a virtual warped dataset.
+
+    Abstracts many details of raster warping and allows access to data
+    that is reprojected as needed.
+
+    This class is backed by an in-memory GDAL VRTWarpedDataset VRT file.
+
+    Attributes
+    ----------
+
+    path : str
+        The file path or network resource identifier for the dataset to
+        be virtually warped.
+    dst_crs : CRS or str
+        The warp operation's destination coordinate reference system.
+    resampling : int
+        One of the values from rasterio.enums.Resampling. The default is
+        `Resampling.nearest`.
+    tolerance : float
+        The maximum error tolerance in input pixels when approximating
+        the warp transformation. The default is 0.125.
+
+    Example
+    -------
+
+    >>> with VirtualWarpedFile('tests/data/RGB.byte.tif', dst_crs='EPSG:3857'
+    ...         ).open() as src:
+    ...     data = src.read()
+
+    """
+    
+    def __init__(self, path, dst_crs=None, resampling=Resampling.nearest,
+                 tolerance=0.125):
+        self.path = path
+        self.dst_crs = dst_crs
+        self.resampling = resampling
+        self.tolerance = tolerance
+
+    @ensure_env
+    def open(self):
```

Comment:
What about just `with VirtualWarpedFile() as src:`?  I don't see a need for an `open()` method since the class itself effectively supplants `rasterio.open()` as a way to access a dataset.  Looks like `@ensure_env` can go on `__enter__()` and still function properly.

```python
def decorator(f):
    print("called")
    return f


class Object(object):

    @decorator
    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        pass


with Object() as o:
    print("inside")


print("before")
o = Object()
print("after")


print("before")
with o as o2:
    print("inside")
print("after")
```